README.md
# terraform-mode.el [![travis badge][travis-badge]][travis-link] [![melpa badge][melpa-badge]][melpa-link] [![melpa stable badge][melpa-stable-badge]][melpa-stable-link]

Major mode of [Terraform](http://www.terraform.io/) configuration file


## Screenshot

![terraform-mode](image/terraform-mode.png)

#### imenu(`helm-imenu`)

![terraform-mode](image/terraform-mode-imenu.png)


## Installation

You can install `terraform-mode.el` from [MELPA](https://melpa.org/) by `package.el`.


## Features

- Syntax highlighting
- Indentation
- imenu


## Customize Variables

#### `terraform-indent-level`(Default: `2`)

Indentation size. You need to call `revert-buffer` if you change this value outer of hook such as `eval-expression`.

## Sample Configuration

```lisp
(custom-set-variables
 '(terraform-indent-level 4))
```

## See Also

- [hcl-mode](https://github.com/syohex/emacs-hcl-mode)

This major-mode inherits from hcl-mode. Most of syntax features, like highlighting, indentation are implemented in hcl-mode.

[travis-badge]: https://travis-ci.org/syohex/emacs-terraform-mode.svg
[travis-link]: https://travis-ci.org/syohex/emacs-terraform-mode
[melpa-link]: https://melpa.org/#/terraform-mode
[melpa-stable-link]: https://stable.melpa.org/#/terraform-mode
[melpa-badge]: https://melpa.org/packages/terraform-mode-badge.svg
[melpa-stable-badge]: https://stable.melpa.org/packages/terraform-mode-badge.svg


Changes
Revision history for terraform-mode.el

Revision 0.06 2017/01/01 syohex
    - Update hcl-mode version for here document highlighting
    - Drop older Emacs support

Revision 0.05 2016/05/03 syohex
    - Update hcl-mode version for here document highlighting

Revision 0.04 2015/12/26 syohex
    - Fix indent size issue(#7, #8)

Revision 0.03 2015/09/01 syohex
    - Change default indent size(Thanks svend)
      2 indent size is recommended by official developpers

Revision 0.02 2015/01/25 syohex
    - Improve for electric-mode

Revision 0.01 2014/08/04 syohex
    - Initial Release


test/test-helper.el
;;; test-helper.el --- helper for testing terraform-mode

;; Copyright (C) 2017 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'terraform-mode)

(defmacro with-terraform-temp-buffer (code &rest body)
  "Insert `code' and enable `terraform-mode'. cursor is beginning of buffer"
  (declare (indent 0) (debug t))
  `(with-temp-buffer
     (insert ,code)
     (goto-char (point-min))
     (terraform-mode)
     (font-lock-fontify-buffer)
     ,@body))

(defun forward-cursor-on (pattern)
  (let ((case-fold-search nil))
    (re-search-forward pattern))
  (goto-char (match-beginning 0)))

(defun backward-cursor-on (pattern)
  (let ((case-fold-search nil))
    (re-search-backward pattern))
  (goto-char (match-beginning 0)))

(defun face-at-cursor-p (face)
  (eq (face-at-point) face))

;;; test-helper.el ends here


test/test-command.el
;;; test-command.el --- Test for terraform-mode commands

;; Copyright (C) 2017 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ert)
(require 'terraform-mode)

(ert-deftest beginning-of-defun ()
  "Move to beginning-of-defun"
  (with-terraform-temp-buffer
    "
variable \"ami\" {
    description = \"the AMI to use\"
}

resource \"aws_instance\" \"web\"{
    ami = ${variable.ami}
    count = 2
}
"

    (forward-cursor-on "use")
    (call-interactively 'hcl-beginning-of-defun)
    (should (looking-at "^variable"))

    (forward-cursor-on "^resource")
    (call-interactively 'hcl-beginning-of-defun)
    (should (looking-at "^variable"))))

(ert-deftest end-of-defun ()
  "Move to end-of-defun"
  (with-terraform-temp-buffer
    "
variable \"ami\" {
    description = \"the AMI to use\"
}
# end1
resource \"aws_instance\" \"web\"{
    ami = ${variable.ami}
    count = 2
}
"

    (forward-cursor-on "use")
    (call-interactively 'hcl-end-of-defun)
    (should (looking-at "^# end1"))

    (forward-cursor-on "^resource")
    (call-interactively 'hcl-end-of-defun)
    (should (eobp))))

;;; test-command.el ends here


test/test-indentation.el
;;; test-indentation.el --- test for indentation

;; Copyright (C) 2017 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ert)
(require 'terraform-mode)

(ert-deftest no-indentation ()
  "No indentation case"
  (with-terraform-temp-buffer
    "
foo = \"val1\"
bar = \"val2\"
"

    (forward-cursor-on "bar")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) 0))))

(ert-deftest no-indentation-with-empty-line ()
  "No indentation case with empty lines"
  (with-terraform-temp-buffer
    "
foo = \"val1\"
    \t
bar = \"val2\"
"

    (forward-cursor-on "bar")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) 0))))

(ert-deftest no-indentation-into-comment ()
  "No indentation case into comment"
  (with-terraform-temp-buffer
    "
    foo = 10
/*
  bar = 20
*/
"

    (forward-cursor-on "bar")
    (let ((cur-indent (current-indentation)))
      (call-interactively 'indent-for-tab-command)
      (should (= (current-indentation) cur-indent)))))

(ert-deftest indentation-into-block ()
  "Indent into block"
  (with-terraform-temp-buffer
    "
provider \"aws\" {
foo = 10
}
"

    (forward-cursor-on "foo")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) terraform-indent-level)))

  (with-terraform-temp-buffer
    "
variable \"aws_amis\" {
foo = 10
}
"

    (forward-cursor-on "foo")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) terraform-indent-level)))

  (with-terraform-temp-buffer
    "
resource \"aws_security_group\" \"group\" {
foo = 10
}
"

    (forward-cursor-on "foo")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) terraform-indent-level))))

(ert-deftest indentation-into-nested-block ()
  "Indent into nested blocks"
  (with-terraform-temp-buffer
    "
resource \"aws_security_group\" \"default\" {
    ingress {
from_port = 22
    }
}
"

    (forward-cursor-on "ingress")
    (let ((cur-indent (current-indentation)))
      (forward-cursor-on "from_port")
      (call-interactively 'indent-for-tab-command)
      (should (= (current-indentation) (+ cur-indent terraform-indent-level))))))

(ert-deftest map-indentation ()
  "Indent for map entry"
  (with-terraform-temp-buffer
    "
map_var {
key = val
}
"

    (forward-cursor-on "key")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) terraform-indent-level))))

(ert-deftest array-indentation ()
  "Indent for array element"
  (with-terraform-temp-buffer
    "
array_var [
\"foo\"
]
"

    (forward-cursor-on "foo")
    (call-interactively 'indent-for-tab-command)
    (should (= (current-indentation) terraform-indent-level))))

(ert-deftest change-indent-size ()
  "Indent for array element"
  (let ((terraform-indent-level 4))
    (with-terraform-temp-buffer
      "
array_var [
\"foo\"
]
"
      (forward-cursor-on "foo")
      (call-interactively 'indent-for-tab-command)
      (should (= (current-indentation) 4)))))

;;; test-indentation.el ends here


terraform-mode.el
;;; terraform-mode.el --- Major mode for terraform configuration file -*- lexical-binding: t -*-

;; Copyright (C) 2017 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>
;; URL: https://github.com/syohex/emacs-terraform-mode
;; Version: 0.06
;; Package-Requires: ((emacs "24.3") (hcl-mode "0.03"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Major mode of terraform configuration file. terraform-mode provides
;; syntax highlighting, indentation function and formatting.

;; Format the current buffer with terraform-format-buffer. To always
;; format terraform buffers when saving, use:
;;   (add-hook 'terraform-mode-hook #'terraform-format-on-save-mode)

;;; Code:

(require 'cl-lib)
(require 'rx)
(require 'hcl-mode)

(defgroup terraform nil
  "Major mode of Terraform configuration file."
  :group 'languages)

(defcustom terraform-indent-level 2
  "The tab width to use when indenting."
  :type 'integer)

(defconst terraform--block-regexp
  "^\\s-*\\(provider\\|resource\\|data\\|module\\|variable\\|output\\)\\s-+\"")

(defconst terraform--atlas-regexp
  "^\\s-*\\(atlas\\)\\s-*")

(defconst terraform--provisioner-regexp
  "^\\s-+\\(provisioner\\)\\s-+\"")

(defvar terraform-font-lock-keywords
  `((,terraform--block-regexp 1 font-lock-function-name-face)
    (,terraform--atlas-regexp 1 font-lock-function-name-face)
    (,terraform--provisioner-regexp 1 font-lock-function-name-face)
    ,@hcl-font-lock-keywords))

(defun terraform-format-buffer ()
  "Rewrite current buffer in a canonical format using terraform fmt."
  (interactive)
  (let ((buf (get-buffer-create "*terraform-fmt*")))
    (if (zerop (call-process-region (point-min) (point-max)
                                    "terraform" nil buf nil "fmt" "-"))
        (let ((point (point))
              (window-start (window-start)))
          (erase-buffer)
          (insert-buffer-substring buf)
          (goto-char point)
          (set-window-start nil window-start))
      (message "terraform fmt: %s" (with-current-buffer buf (buffer-string))))
    (kill-buffer buf)))

(define-minor-mode terraform-format-on-save-mode
  "Run terraform-format-buffer before saving current buffer."
  :lighter ""
  (if terraform-format-on-save-mode
      (add-hook 'before-save-hook #'terraform-format-buffer nil t)
    (remove-hook 'before-save-hook #'terraform-format-buffer t)))

;;;###autoload
(define-derived-mode terraform-mode hcl-mode "Terraform"
  "Major mode for editing terraform configuration file"

  (setq font-lock-defaults '((terraform-font-lock-keywords)))

  ;; indentation
  (make-local-variable 'terraform-indent-level)
  (setq hcl-indent-level terraform-indent-level)

  ;; imenu
  (setq imenu-generic-expression
        '(("resource" "^resource\\s-+\"[^\"]+\"\\s-+\"\\([^\"]+\\)\"" 1)
          ("data" "^data\\s-+\"[^\"]+\"\\s-+\"\\([^\"]+\\)\"" 1)
          ("provider" "^provider\\s-+\"\\([^\"]+\\)\"" 1)
          ("module" "^module\\s-+\"\\([^\"]+\\)\"" 1)
          ("variable" "^variable\\s-+\"\\([^\"]+\\)\"" 1)
          ("output" "^output\\s-+\"\\([^\"]+\\)\"" 1)))
  (imenu-add-to-menubar "Index"))

;;;###autoload
(add-to-list 'auto-mode-alist '("\\.tf\\(vars\\)?\\'" . terraform-mode))

(provide 'terraform-mode)

;;; terraform-mode.el ends here
